client/src/components/Upload/mount.js

```diff
@@ -1,11 +1,29 @@
 import UploadModal from "./UploadModal";
 import { initializeUploadDefaults } from "./config";
 import { mountVueComponent } from "utils/mountVueComponent";
+import { eventHub } from "components/plugins/eventHub";
 
-export function mount(options = {}) {
+// Upload dialog instance
+let uploadVm = null;
+
+export function mountUploadModal(options = {}) {
     const propsData = initializeUploadDefaults(options);
-    const mounter = mountVueComponent(UploadModal);
-    const container = document.createElement("div");
-    document.body.appendChild(container);
-    mounter(propsData, container);
+
```

Comment:
Don't you need to return a promise in this function for 
```
            mountUploadModal(options).then((result) => {
                if (callback) callback(result);
            });
```
to work ?

Reply:
I was going to do that, but it was turning into a mess. I'm actually almost done rewriting the whole Upload component so I just made the callback work again for now.

Reply:
But then you need to drop the promise use in client/src/utils/data.js. Also the select button doesn't appear in the upload modal currently.

Reply:
But to answer your question, yes, you would just make a mount or open function that returns a promise that resolves when an event triggers. The strategy is straightforward and looked something like this:

```js
import { eventHub } from "components/plugins/eventHub";

export function openUploadModal(options) {
    mountUploadModal(options);

    return new Promise((resolve, reject) => {

        eventHub.$once('upload:cancel', (payload) => {
            reject(payload);
        })

        eventHub.$once('upload:close', (payload) => {
            reject(payload);
        })

        eventHub.$once('upload:ok', (payload) => {
            resolve(payload);
        })
    })
}
```

Reply:
But then I have to wade through the component's butchered event handling which led to a testing mess.

Reply:
Eeek, right you are. Forgot to remove all traces of that plan!